github/pkg/adapter/adapter.go

```diff
@@ -19,79 +19,213 @@ package adapter
-	return a, nil
+
+	server := &http.Server{
+		Addr:    ":" + a.port,
+		Handler: a.newRouter(hook),
+	}
+
+	go gracefullShutdown(server, a.logger, stopCh, done)
+
+	a.logger.Infof("Server is ready to handle requests at %s", server.Addr)
+	if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+		return fmt.Errorf("could not listen on %s: %v", server.Addr, err)
+	}
+
+	<-done
+	a.logger.Infof("Server stopped")
+	return nil
+}
+
+func (a *gitHubAdapter) newRouter(hook *gh.Webhook) *http.ServeMux {
+	router := http.NewServeMux()
+	router.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		payload, err := hook.Parse(r, validEvents...)
+		if err != nil {
+			if err == gh.ErrEventNotFound {
+				w.WriteHeader(http.StatusNotFound)
+				a.logger.Info("Event not found")
+				return
+			}
+			w.WriteHeader(http.StatusBadRequest)
+			a.logger.Errorf("Error processing request: %v", err)
+			return
+		}
+		err = a.HandleEvent(payload, r.Header)
+		if err != nil {
+			a.logger.Errorf("Event handler error: %v", err)
+			w.WriteHeader(400)
+			w.Write([]byte(err.Error()))
+			return
+		}
+		a.logger.Infof("Event processed")
+		w.WriteHeader(202)
+		w.Write([]byte("accepted"))
+	})
+	return router
+}
+
+func gracefullShutdown(server *http.Server, logger *zap.SugaredLogger, stopCh <-chan struct{}, done chan<- bool) {
+	<-stopCh
+	logger.Info("Server is shutting down...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	server.SetKeepAlivesEnabled(false)
+	if err := server.Shutdown(ctx); err != nil {
+		logger.Fatalf("Could not gracefully shutdown the server: %v", err)
+	}
+	close(done)
 }
 
 // HandleEvent is invoked whenever an event comes in from GitHub
-func (a *Adapter) HandleEvent(payload interface{}, header http.Header) {
+func (a *gitHubAdapter) HandleEvent(payload interface{}, header http.Header) error {
 	hdr := http.Header(header)
 	err := a.handleEvent(payload, hdr)
-	if err != nil {
-		log.Printf("unexpected error handling GitHub event: %s", err)
-	}
+	return err
 }
 
-func (a *Adapter) handleEvent(payload interface{}, hdr http.Header) error {
-	gitHubEventType := hdr.Get("X-" + GHHeaderEvent)
-	eventID := hdr.Get("X-" + GHHeaderDelivery)
+func (a *gitHubAdapter) handleEvent(payload interface{}, hdr http.Header) error {
+	gitHubEventType := hdr.Get(GHHeaderEvent)
+	if gitHubEventType == "" {
+		return fmt.Errorf("%q header is not set", GHHeaderEvent)
+	}
+	eventID := hdr.Get(GHHeaderDelivery)
+	if eventID == "" {
+		return fmt.Errorf("%q header is not set", GHHeaderDelivery)
+	}
 
-	log.Printf("Handling %s", gitHubEventType)
+	a.logger.Infof("Handling %s", gitHubEventType)
 
 	cloudEventType := sourcesv1alpha1.GitHubEventType(gitHubEventType)
-	subject := subjectFromGitHubEvent(gh.Event(gitHubEventType), payload)
+	subject := subjectFromGitHubEvent(gh.Event(gitHubEventType), payload, a.logger)
 
 	event := cloudevents.NewEvent(cloudevents.VersionV1)
```

Comment:
Can you remove the `VersionV1` in order to use always the latest version? just invoke `NewEvent` without parameters

Reply:
Done.